### PR TITLE
Add partial success for gRPC path.

### PR DESCRIPTION
**Pull request by qingling128**
This PR depends on https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/192.

### CODE REVIEW

**Comment by qingling128** [comment]
Travis is failing because it's installing `google-gax` `v0.8.10` which does not have partial success support yet. I've created a ticket https://github.com/GoogleCloudPlatform/google-cloud-ruby/issues/1799 to ask for a new release of `google-cloud-logging`. `google-gax` version has already been bumped there: https://github.com/GoogleCloudPlatform/google-cloud-ruby/commit/bc749e871757d9e0634b01935dc5039db850fe27#diff-43e7ec6ea7110b556828ce59bd21fd0c

**Comment by qingling128** [comment]
Seems the `google-cloud-logging` gem is waiting for the next `google-gax` release that can go out any day. So we will pin `google-gax` version separately (as in commit https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/174/commits/5bc86d5fd87354e0ff02688a80d0399671cb3adf) for now. Once there is a new `google-cloud-logging` release, we will update accordingly.

**Comment by qingling128** [comment]
Rebased this on `lingshi-grpc-refactor-rebase` and removed the temporary pins of `google-gax` (the issue has been resolved.)

**Comment by igorpeshansky** [comment]
#192 has been merged. You should probably rebase this off `master`.

**Comment by qingling128** [comment]
This is currently blocked by the fact that we have to pin google-cloud-logging to < 1.3 and grpc to < 1.3 to avoid memory leak as in https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/196.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 1]
```diff
@@ -1994,6 +2007,65 @@ def construct_error_details_map(error)
       {}
     end
 
+    # Extract a map of error details from an potentially partially successful
+    # request for gRPC. Return an empty map if @partial_success is not enabled.
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: `gRPC request`.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 2]
```diff
@@ -1890,7 +1903,7 @@ def convert_to_utf8(input)
       end
     end
 
-    # Extract a map of error details from an potentially partially successful
+    # Extract a map of error details from a potentially partially successful
     # request. Return an empty map if @partial_success is not enabled.
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: `REST request`?

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 1 above)
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 3]
```diff
@@ -1927,6 +1944,65 @@ def construct_error_details_map(error)
       {}
     end
 
+    # Extract a map of error details from an potentially partially successful
```
[the remark this answers is not in the document]
Comment (igorpeshansky) [reply lib/fluent/plugin/out_google_cloud.rb]: That makes `request for` redundant, doesn't it?

**Comment by qingling128** [comment]
PTAL. (Seems I could not comment directly on the previous thread. But I've addressed the feedback.)

**Comment by qingling128** [comment]
Rebased off master.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (igorpeshansky) [reply lib/fluent/plugin/out_google_cloud.rb]: Back to `from a potentially partially successful`. :-)

File: lib/fluent/plugin/out_google_cloud.rb [hunk 4]
```diff
@@ -668,7 +659,8 @@ def write_request_via_grpc(entries:,
                   end,
         labels: labels.map do |k, v|
           [k.encode('utf-8'), convert_to_utf8(v)]
-        end.to_h
+        end.to_h,
+        partial_success: @partial_success
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: The comment in line 650 is no longer correct.

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Ah right. Removed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: I... dunno how it happened... It's been a long day.

**Pull request merged**